sycl/test-e2e/Matrix/get_coordinate_ops_impl.hpp

```diff
@@ -0,0 +1,227 @@
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+
+           TResult sum_local_rows[NUM_ROWS] = {0};
+           TResult sum_local_cols[NUM_COLS] = {0};
+
+           if (Use == use::accumulator) {
+             joint_matrix<sub_group, T, use::accumulator, SROWS, SCOLS,
+                          layout::dynamic>
+                 sub;
+
+             joint_matrix_load(
+                 sg, sub,
+                 acc.template get_multi_ptr<access::decorated::no>() +
+                     (sg_startx * SROWS * NUM_COLS) +
+                     sg_starty / sg_size * SCOLS,
+                 NUM_COLS, Layout);
+
+             ext::intel::experimental::matrix::joint_matrix_apply(
+                 sg, sub, [&](T &x, size_t row, size_t col) {
+                   sum_local_rows[row + global_idx * SROWS] += x;
+                   sum_local_cols[col + global_idy / sg_size * SCOLS] += x;
+                 });
+
+           } else {
+             joint_matrix<sub_group, T, Use, SROWS, SCOLS, Layout> sub;
+
+             joint_matrix_load(
+                 sg, sub,
+                 acc.template get_multi_ptr<access::decorated::no>() +
+                     (sg_startx * (SROWS / VF) * NUM_COLS * VF) +
+                     sg_starty / sg_size * SCOLS * VF,
+                 NUM_COLS * VF);
+
+             ext::intel::experimental::matrix::joint_matrix_apply(
+                 sg, sub, [&](T &x, size_t row, size_t col) {
+                   sum_local_rows[row + global_idx * SROWS] += x;
+                   sum_local_cols[col + global_idy / sg_size * SCOLS] += x;
+                 });
+           }
+
+           for (int i = 0; i < NUM_ROWS; i++) {
+             sum_local_rows[i] =
+                 reduce_over_group(sg, sum_local_rows[i], sycl::plus<>());
+             // only Groups leader perform the global reduction
+             if (global_idy % sg_size == 0) {
+               sycl::atomic_ref<TResult, sycl::memory_order::relaxed,
+                                sycl::memory_scope::device>
+                   aref(v_rows[i]);
+               aref.fetch_add(sum_local_rows[i]);
+             }
+           }
+
+           for (int i = 0; i < NUM_COLS; i++) {
+             sum_local_cols[i] =
+                 reduce_over_group(sg, sum_local_cols[i], sycl::plus<>());
+             // only Groups leader perform the global reduction
+             if (global_idy % sg_size == 0) {
+               sycl::atomic_ref<TResult, sycl::memory_order::relaxed,
+                                sycl::memory_scope::device>
+                   aref(v_cols[i]);
+               aref.fetch_add(sum_local_cols[i]);
+             }
+           }
```

Comment:
@YixingZhang007 this code is duplicated. Would be great to outline it to function and reuse.

Reply:
Thanks for the suggestion! I have created a PR for fixing this. The link to the PR is https://github.com/intel/llvm/pull/16429